Sources/PenguinStructures/FixedSizeArray.swift

```diff
@@ -0,0 +1,269 @@
+
+
+/// Statically-sized nonempty collections of homogeneous elements.
+///
+/// This protocol should be thought of as an implementation detail of `ArrayN`;
+/// it is not generally useful.
+public protocol FixedSizeArray : MutableCollection, RandomAccessCollection,
+                          CustomStringConvertible where Index == Int
+{
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.count == c`, where `c` is the capacity of instances.
+  init<Source: Collection>(_ source: Source) where Source.Element == Element
+  // Note: we don't have a generalization to `Sequence` because we couldn't
+  // find an implementation optimizes nearly as well, and in practice
+  // `Sequence`'s that are not `Collection`s are extremely rare.
+
+  /// Creates an instance containing the elements of `source` except the one
+  /// at `targetPosition`.
+  ///
+  /// Requires: `source.indices.contains(targetPosition)`
+  init(_ source: ArrayN<Self>, removingAt targetPosition: Index)
+  
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at `targetPosition`.
+  func inserting(_ newElement: Element, at targetPosition: Index) -> ArrayN<Self>
+}
+
+/// Default implementation of `CustomStringConvertible` conformance.
+public extension FixedSizeArray {
+  var description: String { "\(Array(self))"}
+  
+  @_transparent
+  func withUnsafeBufferPointer<R>(
+      _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafePointer(to: self) { p in
+      try body(
+          UnsafeBufferPointer<Element>(
+              start: UnsafeRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  @_transparent
+  mutating func withUnsafeMutableBufferPointer<R>(
+      _ body: (UnsafeMutableBufferPointer<Element>) throws -> R
+  ) rethrows -> R {
+    let count = self.count
+    return try withUnsafeMutablePointer(to: &self) { p in
+      try body(
+          UnsafeMutableBufferPointer<Element>(
+              start: UnsafeMutableRawPointer(p)
+                  .assumingMemoryBound(to: Element.self),
+              count: count))
+    }
+  }
+
+  /// Returns a fixed-sized collection containing the same elements as `self`,
+  /// with `newElement` inserted at the start.
+  func prepending(_ newElement: Element) -> ArrayN<Self> {
+    .init(head: newElement, tail: self)
+  }
+}
+
+/// A fixed sized collection of 0 elements.
+public struct Array0<T> : FixedSizeArray {
+  public init() {  }
+  
+  /// Creates an instance containing exactly the elements of `source`.
+  ///
+  /// Requires: `source.isEmpty`
+  @inline(__always)
+  public init<Source: Collection>(_ source: Source)
+      where Source.Element == Element
+  {
+    precondition(source.isEmpty, "Too many elements in source")
```

Comment:
For a more helpful error message, could you include the count of remaining elements in source? (Or even maybe the contents of source?)

Reply:
Maybe; I don't want to do anything that would slow this down or bloat the code. IMO the high-order bit for very low-level components like this is not getting a smooth Colab experience, if you know what I mean. And printing the whole contents of the source in particular could be problematic for usability; we don't know how big that might become.  I suggest we file issues for potential improvements like this.